url: introduce native URL class for internal use

jasnell opened the pull request:
Provides a native URL class that uses the WHATWG URL algorithm.
This is intended primarily to support the ES6 modules implementation work (which will need WHATWG URL processing).

/cc @bmeck

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

url

TimothyGu commented:
Can you add some gtests for this new API?

addaleax commented on src/node_url.cc:
```diff
@@ -52,51 +52,20 @@ using v8::Value;
+
+inline void SetFlag(struct url_data* url, enum url_flags flag) {
+  url->flags |= flag;
+}
```
Maybe it’s just me, but I’d probably find it more readable if these methods were just inlined

jasnell replied:
hmm... I guess... I prefer the inlined methods personally but I can change it up

addaleax commented on src/node_url.h:
```diff
@@ -529,6 +529,91 @@ struct url_host {
+                    const size_t len,
+                    enum url_parse_state state_override,
+                    struct url_data* url,
+                    struct url_data* base,
```
This can probably be `const struct url_data* base`?

addaleax commented on src/node_url.h:
```diff
@@ -529,6 +529,91 @@ struct url_host {
+    Parse(input, len, kUnknownState, &context_, nullptr, false);
+  }
+
+  URL(const char* input, const size_t len, URL* base) {
```
Likewise, `const URL* base`?

addaleax commented on src/node_url.h:
```diff
@@ -529,6 +529,91 @@ struct url_host {
+
+  std::string protocol() {
+    return context_.scheme;
+  }
```
I think there’s no reason to not make these methods return const references, i.e. `const std::string& protocol() const { … }` (except `path()`)?

jasnell commented:
@TimothyGu ... yep, I intend to, just ran out of time on Friday and wanted to get this out so that @bmeck had something he could start working with. Will look at adding tests on Monday.

jasnell commented:
@addaleax ... addressed your feedback.

@TimothyGu ... adding a gtest is going to be more involved. I'd like to add that later in a separate PR so that I can get this landed for @bmeck to use.

jasnell commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/6967/

jasnell commented:
CI is 🍏

jasnell commented:
@TimothyGu ... once https://github.com/nodejs/node/pull/11956 lands I will be able to get tests added for the native class.

jasnell commented:
Landed in e26b6c6
